### PR TITLE
Add data processing pipeline for SFT

### PR DESCRIPTION
**Pull request by jialei777**
Add SFT dataset builder and extensive unit tests. Support for
- chat and prompt-completion format
- different loss computing approach
- different truncation logic
- packing

### CODE REVIEW

File: torchprime/data/__init__.py [hunk 1]
```diff
@@ -1,3 +1,11 @@
 """
 The data module contains datasets and efficient dataloading implementations.
 """
+
+from .dataset import make_huggingface_dataset
```
Comment (tengyifei) [inline torchprime/data/__init__.py:5]: Rename this to `make_basic_dataset` or `make_train_dataset` to contrast with `make_sft_dataset`?

File: torchprime/data/dataset.py [hunk 2]
```diff
@@ -1,23 +1,130 @@
...
+
+  data = load_dataset(name, config, split=split, cache_dir=cache_dir)
+  assert isinstance(data, Dataset | DatasetDict)
+  if isinstance(data, DatasetDict):
+    data = data[split]
+  return data
+
+
+def load_hf_or_json_dataset(
+  name: str | None = None,
```
[the remark this answers is not in the document]
Comment (jialei777) [reply torchprime/data/dataset.py:57]: Ok, let me update. I guess it is fine to have different interface as hf datasets package
